core/query/sql.ml

```diff
@@ -150,157 +150,196 @@ let order_by_clause n =
-        let r2 = string_of_base buf3 quote one_table r; Buffer.contents buf3 in
-        buf_add (Arithmetic.gen (l2, op, r2))
+        let ppf3 = Format.formatter_of_buffer buf3 in
+        let l2 = pr_base ppf2 buf2 quote one_table l; Format.pp_print_flush ppf2 (); Buffer.contents buf2 in
+        let r2 = pr_base ppf3 buf3 quote one_table r; Format.pp_print_flush ppf3 (); Buffer.contents buf3 in
+        Format.fprintf ppf "%a" pp_constant (Arithmetic.gen (l2, op, r2))
       | Apply (("intToString" | "stringToInt" | "intToFloat" | "floatToString"
-                | "stringToFloat"), [v]) -> sb v
-      | Apply ("floatToInt", [v]) -> buf_add "floor("; sb v; buf_add ")"
+                | "stringToFloat"), [v]) -> Format.fprintf ppf "%a" pr_b v
+      | Apply ("floatToInt", [v]) -> Format.fprintf ppf "floor(%a)" pr_b v
 
       (* optimisation *)
-      | Apply ("not", [Empty q]) -> buf_add "exists ("; string_of_query buf quote true q; buf_add ")"
-      | Apply ("not", [v]) -> buf_add "not ("; sb v; buf_add ")"
-      | Apply (("negate" | "negatef"), [v]) -> buf_add "-("; sb v; buf_add ")"
-      | Apply ("&&", [v; w]) -> buf_add "("; sb v; buf_add ")"; buf_add " and "; buf_add "("; sb w; buf_add ")"
-      | Apply ("||", [v; w]) -> buf_add "("; sb v; buf_add ")"; buf_add " or "; buf_add "("; sb w; buf_add ")"
-      | Apply ("==", [v; w]) -> buf_add "("; sb v; buf_add ")"; buf_add " = "; buf_add "("; sb w; buf_add ")"
-      | Apply ("<>", [v; w]) -> buf_add "("; sb v; buf_add ")"; buf_add " <> "; buf_add "("; sb w; buf_add ")"
-      | Apply ("<", [v; w]) -> buf_add "("; sb v; buf_add ")"; buf_add " < "; buf_add "("; sb w; buf_add ")"
-      | Apply (">", [v; w]) -> buf_add "("; sb v; buf_add ")"; buf_add " > "; buf_add "("; sb w; buf_add ")"
-      | Apply ("<=", [v; w]) -> buf_add "("; sb v; buf_add ")"; buf_add " <= "; buf_add "("; sb w; buf_add ")"
-      | Apply (">=", [v; w]) -> buf_add "("; sb v; buf_add ")"; buf_add " >= "; buf_add "("; sb w; buf_add ")"
-      | Apply ("RLIKE", [v; w]) -> buf_add "("; sb v; buf_add ")"; buf_add " RLIKE "; buf_add "("; sb w; buf_add ")"
-      | Apply ("LIKE", [v; w]) -> buf_add "("; sb v; buf_add ")"; buf_add " LIKE "; buf_add "("; sb w; buf_add ")"
-      | Apply (f, args) when SqlFuns.is f -> buf_add (SqlFuns.name f); buf_add "("; buf_mapstrcat buf args (fun x -> sb x) ","; buf_add ")"
-      | Apply (f, args) -> buf_add f; buf_add "("; buf_mapstrcat buf args (fun x -> sb x) ","; buf_add ")"
-      | Empty q -> buf_add "not exists ("; string_of_query buf quote true q; buf_add ")"
-      | Length q -> buf_add "select count(*) from ("; string_of_query buf quote true q; buf_add ") as "; buf_add (fresh_dummy_var ())
-      | RowNumber [] -> buf_add "1"
-      | RowNumber ps ->
-        buf_add "row_number() over (order by "; buf_add (String.concat "," (List.map (string_of_projection quote one_table) ps)); buf_add ")"
+      | Apply ("not", [Empty q]) -> Format.fprintf ppf "exists (%a)" pr_q q
```

Comment:
It might be nice to introduce a constant like:
```ocaml
let binary_ops = StringSet.of_list ["&&", "||", "==", ...]
let binary_map op =
  match op with 
  | "&&" -> "and"
  | "||" -> "or"
  | "==" -> "="
  | _ -> ""
```

And then in your code you can just have:
```ocaml
 | Apply (op, [v; w]) when StringSet.mem op binary_ops -> Format.fprintf ppf "(%a) %s (%a)" pr_b v (binary_map op) pr_b w
```

It might also be nice to do the same with unary operators.